ssdp/ssdp.go

```diff
@@ -71,7 +80,66 @@ func SSDPRawSearchCtx(
 	if err != nil {
 		return nil, err
 	}
+	return processSSDPResponses(searchTarget, allResponses)
+}
+
+// RawSearch performs a fairly raw SSDP search request, and returns the
+// unique response(s) that it receives. Each response has the requested
+// searchTarget, a USN, and a valid location. If the provided context times out
+// or is canceled, the search will be aborted. numSends is the number of
+// requests to send - 3 is a reasonable value for this.
+func RawSearch(
+	ctx context.Context,
+	httpu HTTPUClientCtx,
+	searchTarget string,
+	numSends int,
+) ([]*http.Response, error) {
+	// We need a timeout value to include in the SSDP request; get it by
+	// checking the deadline on the context.
+	var maxWaitSeconds int
+	if deadline, ok := ctx.Deadline(); ok {
+		maxWaitSeconds = int(deadline.Sub(time.Now()) / time.Second)
+	} else {
+		// Pick a default timeout of 3 seconds if none was provided.
+		maxWaitSeconds = 3
+
+		var cancel func()
+		ctx, cancel = context.WithTimeout(ctx, time.Duration(maxWaitSeconds)*time.Second)
+		defer cancel()
+	}
+
+	// Check the timeout on the context (if any); if the context would time
```

Comment:
This line through and including the request creation (lines 111-130) look like they are common code that can be factored out (`func prepareRequest(...) (*http.Request, error)` ?).

Or could `SSDPRawSearchCtx` be implemented mostly by calling `RawSearch`?

Reply:
I factored them out into a shared function. I didn't want to change the behaviour of `SSDPRawSearchCtx` since users may have been expecting that the `Do` function would be called, and `RawSearch` calls `DoWithContext` instead (and requires a different interface type).